### PR TITLE
Delay zero crossing and event handling

### LINKED ISSUE #7536: delay not working correctly
We're going to implement test models for `Modelica.Electrical.Analog.Lines.TLine` which rely on the delay-operator.
I'm expecting that the output is an exact copy of the input but delayTime later, but in OM it isn't:
The slopes (at the beginning and the end of a pulse) are finite (instead of infinite).
Pls. have a look at the enclosed example. 
OM won't be able to reproduce the comaprison signals of the examples.

OM v1.17.0 (64 bit) on Windows 10 (64 bit) pro
[TestDelay.zip](https://github.com/OpenModelica/OpenModelica/files/6625615/TestDelay.zip)

### LINKED ISSUE #7792: Delay operator not working as expected when dealing with discontinuous equations
### Description
**Delay** operator **not working as expected** when dealing with **discontinuous equations**

### Steps to reproduce
take any sort of discontinuous equation and assign the delayed version of that equation to a new variable. 
Observation: While the original equation is discontinuous, the delay function isn't. 

We really 💖 minimal working examples. With these reproducing a bug is straightforward and we can respond faster.

Example minimalExample.mos:

```Modelica
model Delay_Bug
  Real x;
  Real y;
equation
  x  = if time<0.5 then 1 else time;
  y = delay(x,0.1);
end Delay_Bug;
```

### Expected behavior
The discontinuity of x at time=0.5 is expected to be caught by the delay function at time=0.6 which isn't happening. Y is continuous, despite x being discontinuous
### Screenshots
image

### Version and OS
Connected to OpenModelica v1.17.0 (64-bit)
Connected to OMSimulator v2.0.0.post284-gc8ec782-mingw
Windows10

### PR DESCRIPTION
**Pull request by AnHeuermann**
### Related Issues

Fixes #7536 and fixes #7792

### Purpose

This is work in progress. The delay operator should be able to delay events and trigger them at the correct time.

#### TODO

- [x] Fix `findTime` function. It is finding events to early.
- [x] Remove debug prints / clean up.
- [x] Add new or validate existing test cases with delay and events.
- [x] Update documentation(?)

### Approach

- Added new debug flag `LOG_DELAY`
- In `storeDelayedExpression` events are added to `RINGBUFFER*  delayStruct`. When `storeDelayedExpression` stored events should be noticed. An event iteration should come next.
- Added zero-crossing function for the `delay` operator

### CODE REVIEW

**Comment by AnHeuermann** [comment]
```modelica
model Delay_Bug
  Real x;
  Real y;
equation
  x = if time<0.5 then 1 else time;
  y = delay(x,0.1);
  annotation(
    experiment(StartTime = 0, StopTime = 1, Tolerance = 1e-6, Interval = 0.05));
end Delay_Bug;
```

image

There should be sharp event at time `t=0.6`, but none is triggered.

**Comment by AnHeuermann** [comment]
After way to much debugging I should now have something that is triggering events at the correct location:

For the example provided in https://github.com/OpenModelica/OpenModelica/issues/7536#issue-916428332:
image

For the example provided in https://github.com/OpenModelica/OpenModelica/issues/7536#issuecomment-858118413:
image
This one is not perfect but could be improved with a better interpolation/ extrapolation method. Currently we only use a linear one. But the events should be sharp.

I hope I won't get to much failing tests...

**Comment by AnHeuermann** [comment]
I'm not really surprised that a model like
```modelica
model nls10
  Real x(start = 1/2, fixed = true);
  Real y;
equation
  der(x) = delay(y,0.01);
  (if(x < 0) then 2*y -10 else 3*y + 1) = (if(x < 0.5) then 3*x*y+17 else x*y/2 - 13);
end nls10;
```

has event shattering. But in that case we simply had to many tiny steps saved in the delay buffer and where finding "events" everywhere. I made it less strict now.

**Comment by AnHeuermann** [comment]
I still have some issue with examples like:
```modelica
model Delay
  Real x;
  Real y;
  Real z;
  Real w;
equation
  x = delay(y+1, 2.5);
  y = if time < 5 then sin(time) else cos(time);
  z = delay(y, 0.01);
  w = delay(delay(y, 0.5), 1.5);
end Delay;
```

I'm finding one event to many (at `time=5.01`)
```
// LOG_EVENTS        | info    | state event at time=5.000000006
// |                 | |       | | [2] time < 5.0
// LOG_EVENTS        | info    | state event at time=5.010000006
// |                 | |       | | [3] delayZeroCrossing(1, 2, y, 0.01, 0.01) > 0.0
// LOG_EVENTS        | info    | state event at time=7.500000006
// |                 | |       | | [1] delayZeroCrossing(0, 0, y, 2.5, 2.5) > 0.0
// LOG_SUCCESS       | info    | The simulation finished successfully.
```

But I don't think that is a serious issue for the time being. This can be fixed later (a.k.a. never).

File: OMCompiler/SimulationRuntime/c/simulation/solver/delay.c [hunk 1]
```diff
@@ -41,138 +41,197 @@
...
       value1 = exprValue;
-      infoStreamPrint(LOG_EVENTS_V, 0, "delayImpl: times %g and %g", time0, time1);
-      infoStreamPrint(LOG_EVENTS_V, 0, "delayImpl: values %g and  %g", value0, value1);
     }
     else
     {
-      i = findTime(timeStamp, delayStruct);
+      i = findTime(timeStamp, delayStruct, &foundEvent);
+      if (foundEvent) {
+        // TODO: Do something if an event was found!
```
Comment (kabdelhak) [inline OMCompiler/SimulationRuntime/c/simulation/solver/delay.c:232]: Is this finished or did you forget this? :)

File: OMCompiler/SimulationRuntime/c/simulation/solver/delay.c [hunk 2]
```diff
@@ -185,27 +244,85 @@ double delayImpl(DATA* data, threadData_t *threadData, int exprNumber, double ex
...
+
+  if (ringBufferLength(delayStruct) == 0)
+  {
+    return zeroCrossingValue;
+  }
+
+  /* Find first event on ring buffer */
+  findTime(time - delayTime, delayStruct, &foundEvent);
+
+  /* Flip sign of ZC if an event was found */
```
Comment (kabdelhak) [inline OMCompiler/SimulationRuntime/c/simulation/solver/delay.c:300]: why not just return 1 and -1? does this work if the zeroCrossingValue is 0? or is it always just 1 and -1?

**Review by kabdelhak — APPROVED** [review]
LGTM, just added some small questions

File: OMCompiler/SimulationRuntime/c/simulation/solver/delay.c (hunk 1 above)
Comment (AnHeuermann) [reply OMCompiler/SimulationRuntime/c/simulation/solver/delay.c:232]: Not really sure. I guess in that case I should just remove it.

**Comment by casella** [comment]
No visible adverse effects in the [library testing regression report](https://libraries.openmodelica.org/branches/history/master/2021-11-18%2022:44:34..2021-11-23%2015:44:35.html), which is unfortunately a bit crowded since the testing job was stuck for a few days in a row.

**Pull request merged**